<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="datelife_use gets secondary calibrations available for any
pair of given taxon names, mined from the opentree_chronograms object,
congruifies them, and uses them to date a given tree topology with the
algorithm defined in dating_method. If no tree topology is provided,
it will attempt to get one for the given taxon names from Open Tree of Life
synthetic tree, using make_bold_otol_tree()."><title>Generate one or multiple chronograms for a set of given taxon names. — datelife_use • datelife</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate one or multiple chronograms for a set of given taxon names. — datelife_use"><meta property="og:description" content="datelife_use gets secondary calibrations available for any
pair of given taxon names, mined from the opentree_chronograms object,
congruifies them, and uses them to date a given tree topology with the
algorithm defined in dating_method. If no tree topology is provided,
it will attempt to get one for the given taxon names from Open Tree of Life
synthetic tree, using make_bold_otol_tree()."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">datelife</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.9</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Getting_started_with_datelife.html">Getting started</a>
    <a class="dropdown-item" href="../articles/fringiliidae.html">Case study: the true finches</a>
    <a class="dropdown-item" href="../articles/make_datelife_query.html">Making a DateLife query</a>
    <a class="dropdown-item" href="../articles/making_bold_trees.html">Estimating initial branch lengths</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/phylotastic/datelife/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Generate one or multiple chronograms for a set of given taxon names.</h1>
      <small class="dont-index">Source: <a href="https://github.com/phylotastic/datelife/blob/HEAD/R/datelife_use.R" class="external-link"><code>R/datelife_use.R</code></a></small>
      <div class="d-none name"><code>datelife_use.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>datelife_use</code> gets secondary calibrations available for any
pair of given taxon names, mined from the <a href="opentree_chronograms.html">opentree_chronograms</a> object,
congruifies them, and uses them to date a given tree topology with the
algorithm defined in <code>dating_method</code>. If no tree topology is provided,
it will attempt to get one for the given taxon names from Open Tree of Life
synthetic tree, using <code><a href="make_bold_otol_tree.html">make_bold_otol_tree()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">datelife_use</span><span class="op">(</span>input <span class="op">=</span> <span class="cn">NULL</span>, each <span class="op">=</span> <span class="cn">FALSE</span>, dating_method <span class="op">=</span> <span class="st">"bladj"</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>One of the following:</p><dl><dt>A character vector</dt>
<dd><p>With taxon names as a single comma separated
starting or concatenated with <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code>.</p></dd>

<dt>A phylogenetic tree with taxon names as tip labels</dt>
<dd><p>As a <code>phylo</code> or
<code>multiPhylo</code> object, OR as a newick character string.</p></dd>

<dt>A <code>datelifeQuery</code> object</dt>
<dd><p>An output from <code><a href="make_datelife_query.html">make_datelife_query()</a></code>.</p></dd>


</dl></dd>


<dt>each</dt>
<dd><p>Boolean, default to <code>FALSE</code>: all calibrations are returned in
the same <code>data.frame</code>. If <code>TRUE</code>, calibrations from each chronogram are returned
in separate data frames.</p></dd>


<dt>dating_method</dt>
<dd><p>Tree dating algorithm to use. Options are "bladj" or "pathd8"
(Webb et al., 2008, <a href="https://doi.org/10.1093/bioinformatics/btn358" class="external-link">doi:10.1093/bioinformatics/btn358</a>
; Britton et al., 2007,
<a href="https://doi.org/10.1080/10635150701613783" class="external-link">doi:10.1080/10635150701613783</a>
).</p></dd>


<dt>...</dt>
<dd><p>Arguments passed on to <code><a href="make_datelife_query.html">make_datelife_query</a></code></p><dl><dt><code>use_tnrs</code></dt>
<dd><p>Whether to use Open Tree of Life's Taxonomic Name Resolution Service (TNRS)
to process input taxon names. Default to <code>TRUE</code>, it corrects misspellings and
taxonomic name variations with <code><a href="tnrs_match.html">tnrs_match()</a></code>, a wrapper of <code><a href="https://docs.ropensci.org/rotl/reference/tnrs_match_names.html" class="external-link">rotl::tnrs_match_names()</a></code>.</p></dd>

    <dt><code>get_spp_from_taxon</code></dt>
<dd><p>Whether to search ages for all species belonging to a
given taxon or not. Default to <code>FALSE</code>. If <code>TRUE</code>, it must have same length as input.
If input is a newick string with some clades it will be converted to a <code>phylo</code>
object, and the order of <code>get_spp_from_taxon</code> will match <code>phy$tip.label</code>.</p></dd>

    <dt><code>reference_taxonomy</code></dt>
<dd><p>A character vector specifying the reference taxonomy to use for TNRS.
Options are "ott", "ncbi", "gbif" or "irmng". The function defaults to "ott".</p></dd>

  
</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code>phylo</code> or <code>multiPhylo</code> object with branch lengths proportional to time.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>If <code>input</code> is a vector of taxon names, the function will attempt to reconstruct a BOLD
tree with <code><a href="make_bold_otol_tree.html">make_bold_otol_tree()</a></code> to get a tree with branch lengths. If it fails,
it will get an Open Tree of Life synthetic tree topology.
The function then calls <code><a href="use_calibrations.html">use_calibrations()</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="more">More<a class="anchor" aria-label="anchor" href="#more"></a></h2>
    <p>The output object stores the used <code>calibrations</code> and <code>dating_method</code> as
<code>attributes(output)$datelife_calibrations</code> and <code>attributes(output)$dating_method</code>.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Brian O’Meara, Jonathan Eastman, Tracy Heath, April Wright, Klaus Schliep, Scott Chamberlain, Peter Midford, Luke Harmon, Joseph Brown, Matt Pennell, Mike Alfaro, Luna L. Sanchez Reyes.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

